File: frontend/src/api/testSuite.ts
import { apiClient } from '@/utils/api'
import type { PaginationResponse } from '@/types'

export interface TestSuite {
  id: string
  planId: string
  name: string
  description?: string
  gitEnabled?: string  // Git配置是否启用: 'true'/'false'
  gitRepoUrl?: string  // 可选
  gitBranch?: string  // 可选
  gitToken?: string  // 可选
  environmentId: string
  executionCommand: string
  caseIds: string[]
  status: 'pending' | 'running' | 'completed' | 'failed'
  createdAt: string
  updatedAt: string
}

export interface TestSuiteExecution {
  id: string
  suiteId: string
  caseId: string
  caseName?: string
  environmentId: string
  environmentName?: string
  executorId: string
  result: 'passed' | 'failed' | 'error' | 'skipped'
  duration?: string
  logOutput?: string
  errorMessage?: string
  executedAt: string
  createdAt: string
}

export interface TestSuiteCreate {
  name: string
  description?: string
  gitEnabled?: string  // Git配置是否启用: 'true'/'false'
  gitRepoUrl?: string  // 可选
  gitBranch?: string  // 可选
  gitToken?: string  // 可选
  environmentId: string
  executionCommand: string
  caseIds: string[]
}

export const testSuiteApi = {
  // 获取测试套列表
  getTestSuites: async (planId: string, params?: { skip?: number; limit?: number }): Promise<PaginationResponse<TestSuite>> => {
    const queryParams = new URLSearchParams()
    if (params) {
      Object.entries(params).forEach(([key, value]) => {
        if (value !== undefined && value !== null) {
          queryParams.append(key, String(value))
        }
      })
    }
    const url = `/test-plans/${planId}/suites${queryParams.toString() ? '?' + queryParams.toString() : ''}`
    return apiClient.get(url)
  },

  // 获取测试套详情
  getTestSuite: async (suiteId: string): Promise<TestSuite> => {
    return apiClient.get(`/test-plans/suites/${suiteId}`)
  },

  // 创建测试套
  createTestSuite: async (planId: string, data: TestSuiteCreate): Promise<TestSuite> => {
    return apiClient.post(`/test-plans/${planId}/suites`, data)
  },

  // 更新测试套
  updateTestSuite: async (suiteId: string, data: Partial<TestSuiteCreate>): Promise<TestSuite> => {
    return apiClient.put(`/test-plans/suites/${suiteId}`, data)
  },

  // 删除测试套
  deleteTestSuite: async (suiteId: string): Promise<void> => {
    return apiClient.delete(`/test-plans/suites/${suiteId}`)
  },

  // 执行测试套
  executeTestSuite: async (suiteId: string): Promise<TestSuite> => {
    return apiClient.post(`/test-plans/suites/${suiteId}/execute`)
  },

  // 取消测试套执行
  cancelTestSuite: async (suiteId: string): Promise<TestSuite> => {
    return apiClient.post(`/test-plans/suites/${suiteId}/cancel`)
  },

  // 获取测试套执行记录
  getSuiteExecutions: async (suiteId: string, params?: { skip?: number; limit?: number }): Promise<PaginationResponse<TestSuiteExecution>> => {
    const queryParams = new URLSearchParams()
    if (params) {
      Object.entries(params).forEach(([key, value]) => {
        if (value !== undefined && value !== null) {
          queryParams.append(key, String(value))
        }
      })
    }
    const url = `/test-plans/suites/${suiteId}/executions${queryParams.toString() ? '?' + queryParams.toString() : ''}`
    return apiClient.get(url)
  },

  // 获取测试套日志
  getSuiteLogs: async (suiteId: string, params?: { skip?: number; limit?: number; executionId?: string; logId?: string }): Promise<PaginationResponse<{ id: string; level?: string; message: string; timestamp: string; createdAt: string; execution_id?: string }>> => {
    const queryParams = new URLSearchParams()
    if (params) {
      Object.entries(params).forEach(([key, value]) => {
        if (value !== undefined && value !== null) {
          queryParams.append(key, String(value))
        }
      })
    }
    const url = `/test-plans/suites/${suiteId}/logs${queryParams.toString() ? '?' + queryParams.toString() : ''}`
    return apiClient.get(url)
  },

  // 获取测试套执行历史（按execution_id分组）
  getSuiteSuiteExecutions: async (
    suiteId: string,
    params?: {
      skip?: number;
      limit?: number;
      search?: string;
      result?: string;
      startDate?: string;
      endDate?: string;
    }
  ): Promise<PaginationResponse<any>> => {
    const queryParams = new URLSearchParams()
    if (params) {
      Object.entries(params).forEach(([key, value]) => {
        if (value !== undefined && value !== null) {
          queryParams.append(key, String(value))
        }
      })
    }
    const url = `/test-plans/suites/${suiteId}/suite-executions${queryParams.toString() ? '?' + queryParams.toString() : ''}`
    return apiClient.get(url)
  },

  // 删除测试套执行历史
  deleteSuiteExecution: async (suiteId: string, executionId: string): Promise<void> => {
    return apiClient.delete(`/test-plans/suites/${suiteId}/suite-executions/${executionId}`)
  }
}

File: backend/schemas/test_suite.py
"""测试套相关模式"""
from pydantic import BaseModel, field_validator
from typing import Optional, List
from uuid import UUID
from datetime import datetime


class TestSuiteBase(BaseModel):
    """测试套基础模式"""
    name: str
    description: Optional[str] = None
    git_enabled: Optional[str] = "false"  # Git配置是否启用: true/false
    git_repo_url: Optional[str] = None  # 可选
    git_branch: Optional[str] = "main"  # 可选
    git_token: Optional[str] = None  # 可选
    environment_id: str
    execution_command: str
    case_ids: List[str]


class TestSuiteCreate(TestSuiteBase):
    """创建测试套请求"""
    plan_id: str


class TestSuiteUpdate(BaseModel):
    """更新测试套请求"""
    name: Optional[str] = None
    description: Optional[str] = None
    git_enabled: Optional[str] = None  # Git配置是否启用: true/false
    git_repo_url: Optional[str] = None
    git_branch: Optional[str] = None
    git_token: Optional[str] = None
    environment_id: Optional[str] = None
    execution_command: Optional[str] = None
    case_ids: Optional[List[str]] = None
    status: Optional[str] = None
    plan_id: Optional[str] = None  # 允许修改关联的测试计划


class TestSuiteResponse(TestSuiteBase):
    """测试套响应"""
    id: str
    plan_id: str
    status: str
    created_at: datetime
    updated_at: datetime
    created_by: Optional[str] = None
    updated_by: Optional[str] = None
    
    class Config:
        from_attributes = True


class TestSuiteExecutionResponse(BaseModel):
    """测试套执行记录响应"""
    id: str
    suite_id: str
    case_id: str
    case_name: Optional[str] = None
    environment_id: str
    environment_name: Optional[str] = None
    executor_id: str
    result: str
    duration: Optional[str] = None
    log_output: Optional[str] = None
    error_message: Optional[str] = None
    executed_at: datetime
    created_at: datetime
    
    class Config:
        from_attributes = True


class TestSuiteCancelRequest(BaseModel):
    """取消测试套执行请求"""
    executionId: Optional[str] = None
    
    class Config:
        populate_by_name = True

